lms/static/js/student_account/multiple_enterprise.js

```diff
@@ -15,35 +16,79 @@
             /**
              * Fetch the learner data, then redirect the user to a enterprise selection page if multiple
              * enterprises were found.
-             * @param  {string} nextUrl The URL to redirect to after multiple enterprise selection.
+             * @param  {string} nextUrl The URL to redirect to after multiple enterprise selection or incase
+             * the selection page is bypassed e.g. when dealing with direct enrolment urls.
              */
             check: function(nextUrl) {
-                var redirectUrl = this.urls.multipleEnterpriseUrl + encodeURIComponent(nextUrl);
+                var view = this;
+                var selectionPageUrl = this.urls.multipleEnterpriseUrl + encodeURIComponent(nextUrl);
                 var username = Utils.userFromEdxUserCookie().username;
                 var next = nextUrl || '/';
+                var enterpriseInUrl = this.getEnterpriseFromUrl(nextUrl);
+                var userInEnterprise = false;
+                var userWithMultipleEnterprises = false;
                 $.ajax({
                     url: this.urls.learners + '?username=' + username,
                     type: 'GET',
                     contentType: 'application/json; charset=utf-8',
                     headers: this.headers,
                     context: this
                 }).fail(function() {
-                    this.redirect(next);
+                    view.redirect(next);
                 }).done(function(response) {
-                    if (response.count > 1 && redirectUrl) {
-                        this.redirect(redirectUrl);
+                    userWithMultipleEnterprises = (response.count > 1);
+                    if (userWithMultipleEnterprises) {
+                        if (enterpriseInUrl) {
+                            userInEnterprise = view.checkEnterpriseExists(response, enterpriseInUrl);
+                            if (userInEnterprise) {
+                                view.activate(enterpriseInUrl).fail(function() {
+                                    view.redirect(selectionPageUrl);
+                                }).done(function() {
+                                    view.redirect(next);
+                                });
+                            } else {
+                                view.redirect(selectionPageUrl);
+                            }
+                        } else {
+                            view.redirect(selectionPageUrl);
+                        }
                     } else {
-                        this.redirect(next);
+                        view.redirect(next);
```

Comment:
can we optimize these multiple if...else blocks?

Reply:
Done